lib/galaxy/objectstore/onedata.py

```diff
@@ -0,0 +1,557 @@
+            if not os.path.exists(cache_dir):
+                os.makedirs(cache_dir, exist_ok=True)
+
+            if not dir_only:
+                rel_path = os.path.join(rel_path, alt_name if alt_name else f"dataset_{self._get_object_id(obj)}.dat")
+                # need this line to set the dataset filename, not sure how this is done - filesystem is monitored?
+                open(os.path.join(self.staging_path, rel_path), "w").close()
+                self._client.create_file(self.space_name, rel_path, 'REG', create_parents=True)
+        return self
+
+    def _empty(self, obj, **kwargs):
+        if self._exists(obj, **kwargs):
+            return bool(self._size(obj, **kwargs) > 0)
+        else:
+            raise ObjectNotFound(f"objectstore.empty, object does not exist: {obj}, kwargs: {kwargs}")
+
+    def _size(self, obj, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        if self._in_cache(rel_path):
+            try:
+                return self._get_size_in_cache(rel_path)
+            except OSError as ex:
+                log.info("Could not get size of file '%s' in local cache, will try Onedata. Error: %s", rel_path, ex)
+        elif self._exists(obj, **kwargs):
+            return self._get_size_in_onedata(rel_path)
+        log.warning("Did not find dataset '%s', returning 0 for size", rel_path)
+        return 0
+
+    def _delete(self, obj, entire_dir=False, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        extra_dir = kwargs.get("extra_dir", None)
+        base_dir = kwargs.get("base_dir", None)
+        dir_only = kwargs.get("dir_only", False)
+        obj_dir = kwargs.get("obj_dir", False)
+
+        try:
+            # Remove temporary data in JOB_WORK directory
+            if base_dir and dir_only and obj_dir:
+                shutil.rmtree(os.path.abspath(rel_path))
+                return True
+
+            # Delete from cache first
+            if entire_dir and extra_dir:
+                shutil.rmtree(self._get_cache_path(rel_path), ignore_errors=True)
+            else:
+                unlink(self._get_cache_path(rel_path), ignore_errors=True)
+
+            # Delete from Onedata as well
+            if self._exists_in_onedata(rel_path):
+                self._client.remove(self.space_name, rel_path)
+                return True
+        except OnedataRESTError:
+            log.exception("Could not delete '%s' from Onedata", rel_path)
+        except OSError:
+            log.exception("%s delete error", self._get_filename(obj, **kwargs))
+        return False
+
+    def _get_data(self, obj, start=0, count=-1, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        # Check cache first and get file if not there
+        if not self._in_cache(rel_path) or self._get_size_in_cache(rel_path) == 0:
+            self._pull_into_cache(rel_path)
+        # Read the file content from cache
+        # TODO with open?
```

Comment:
to twoje, czy skopiowane skądś? patrzę np. na irods i tam nie ma

ale pewnie to była myśl aby używać `with` by się ładnie plk zamknął, chyba warto?